### PR TITLE
fix(world-model): build default discounts in float32 regardless of reward dtype

### LINKED ISSUE #386: World-model learning loops build the default discount with jnp.full_like(rewards, gamma), so integer/bool rewards train the discount head on 0 or 1
## Summary

`run_action_conditioned_world_model_learning_loop` (`alberta_framework/core/world_model.py`) and `run_latent_world_model_learning_loop` (`alberta_framework/core/latent_world_model.py`) document `discounts=None` as "use `model.config.gamma`" and build that default with `jnp.full_like(rewards, gamma)`. `full_like` inherits the **reward array's dtype**, so `gamma` is cast into it: an `int32` reward array (binary/counting rewards, `(arange % 2).astype(int32)`) yields discount targets of `0`, a `bool` reward array yields `True` (= 1.0, above the `[0, gamma]` prediction clip and unreachable). Nothing rejects a non-float reward array — the per-step `update()` casts rewards to float32 itself, every transaction commits, `updates_applied` is all-`True`, and `discount_errors` come back small and plausible. The discount head — which `dreaming.py` uses to terminate rollouts — is silently trained to predict immediate termination.

## Falsifier (deterministic, pristine `main` @ `4f75cc8`)

```python
import jax.numpy as jnp, jax.random as jr
from alberta_framework.core.world_model import (
    ActionConditionedWorldModel, ActionConditionedWorldModelConfig,
    run_action_conditioned_world_model_learning_loop as run)
cfg = ActionConditionedWorldModelConfig(observation_dim=2, n_actions=2, hidden_sizes=(), step_size=0.05, sparsity=0.0)
model = ActionConditionedWorldModel(cfg); state = model.init(jr.key(8))
obs = jnp.array([[0., 0.], [.1, 0.], [.1, .2]]); nxt = jnp.array([[.1, 0.], [.1, .2], [.2, .2]]); acts = jnp.array([0, 1, 0])
print("jnp.full_like(int32 rewards, 0.99) ->", jnp.full_like(jnp.array([1, 0, 1]), 0.99))
for rewards in (jnp.array([1., 0., 1.]), jnp.array([1, 0, 1]), jnp.array([True, False, True])):
    r = run(model, state, obs, acts, rewards, nxt)
    print(rewards.dtype, "discount_errors:", r.discount_errors, "applied:", bool(jnp.all(r.updates_applied)))
```

Output on `main`:

```
jnp.full_like(int32 rewards, 0.99) -> [0 0 0]
float32 discount_errors: [ 0.99  0.9899 0.9895] applied: True
int32   discount_errors: [ 0.     -0.0001 -0.0005] applied: True
bool    discount_errors: [ 1.      0.9999  0.9995] applied: True
```

Same shape in the latent loop (`discount predictions [0.08 0.00 0.08]` for int32 rewards vs `[0.73 0.63 0.76]` for float32).

Expected: the default discount vector is built in float32 irrespective of the reward dtype (`jnp.full(shape, gamma, dtype=float32)`), so integer/bool rewards train the same discount target as float rewards.

## Dedup

`gh issue/pr list --search "full_like" / "default discount" / "world model discount"`: nothing on this surface (#168/#169 is GVFSpec discount validation, a different mechanism). No open PR touches `core/world_model.py` or `core/latent_world_model.py`.

Fix PR follows (failing-test-first).

### PR DESCRIPTION
**Pull request by Prabhat1308**
Closes #386.

## Issue

Both world-model learning loops derived the default discount vector with `jnp.full_like(rewards, gamma)`, which inherits the reward dtype: `int32` rewards truncated `gamma` to `0` and `bool` rewards to `True`, training the discount head on the wrong constant while every update was accepted (`discount_errors ≈ 0` instead of `≈ 0.99`).

## New state

The default is built with `jnp.full(jnp.shape(rewards), gamma, dtype=jnp.float32)` in both loops; integer, bool, and float reward arrays now produce identical discount targets, predictions, and errors. Explicitly passed `discounts` are unchanged.

Failing-test-first: `test_default_discounts_do_not_inherit_the_reward_dtype` (action-conditioned; float/int32/bool vs an explicit `full(gamma)` run) and `test_latent_default_discounts_do_not_inherit_the_reward_dtype` fail on `main` and pass here.

## Evidence

Falsifier from #386 at this head:

```text
ACWM int32 rewards: discount targets = [0.99 0.99 0.99 0.99] | updates_applied all = True
LWM  int32 rewards: discount predictions = [0.7348 0.6286 0.7645] (== float32 run)
```

Verification at `486a46c0c8cfbd60c828965ac316639f689add4a`:

```text
.venv/bin/python -m pytest tests/test_action_conditioned_world_model.py tests/test_latent_world_model.py tests/test_world_model*.py -q -o addopts=""   -> 66 passed
.venv/bin/python -m ruff check .                                                                                                    -> All checks passed!
.venv/bin/python -m mypy                                                                                                            -> Success: no issues found in 164 source files
```

Neither module is a registered evidence source; no benchmark lane, seeds, or `outputs/` artifacts were touched; no `CHANGELOG.md` edit (maintainer-recorded).

CLI sessions cannot mint GitHub attachment URLs, so the run output above is inline rather than attached.

- [x] logs: inline above (focused pytest, ruff, mypy, falsifier output at the evidence head)

Provider/model/client: anthropic / claude-fable-5 / Claude Code 2.1.233 (contribute-to-asi skill revision 72e4239e).

### CODE REVIEW

**Review by lalalune — APPROVED** [review]
Independent review at exact head 486a46c: READY. Both default-discount learning loops now construct float32 targets from reward shape rather than inheriting integer/bool reward dtype; explicit discounts remain untouched. Verified in a clean current-main composition with actor-critic changes: 244 focused/adjacent actor/world-model/ensemble tests passed, Ruff and strict mypy passed, and diff hygiene is clean. Float, integer, and boolean reward arrays produce identical default discount predictions/errors. No registered evidence source or output artifact changes.
